File: sdks/js_sdk/src/secrets/backends/keyring.ts
import type { SecretsBackend } from '../backend.js'
import { SecretBackendError, SecretNotFound } from '../errors.js'
import { Scope, SecretKey, SecretValue } from '../types.js'

export class KeyringBackend implements SecretsBackend {
  private keytar: any
  private servicePrefix: string

  constructor(servicePrefix = 'sw4rm') {
    this.servicePrefix = servicePrefix
  }

  private async ensureKeytar() {
    if (!this.keytar) {
      try {
        const mod = await import('keytar')
        this.keytar = mod.default ?? mod
      } catch (e: any) {
        throw new SecretBackendError("keyring backend unavailable; install 'keytar' package or use file backend")
      }
    }
  }

  private service(scope: Scope) { return `${this.servicePrefix}:${scope.label()}` }

  async set(scope: Scope, key: SecretKey, value: SecretValue): Promise<void> {
    await this.ensureKeytar()
    await this.keytar.setPassword(this.service(scope), key.key, value.value)
  }

  async get(scope: Scope, key: SecretKey): Promise<string> {
    await this.ensureKeytar()
    const v = await this.keytar.getPassword(this.service(scope), key.key)
    if (v == null) throw new SecretNotFound(scope.label(), key.key)
    return v
  }

  async list(): Promise<Map<[string | null, string], string>> {
    // Keytar does not support listing safely; return empty.
    return new Map()
  }
}

File: sdks/rust_sdk/src/secrets/backends/keyring.rs
#[cfg(feature = "keyring")]
use std::collections::HashMap;

#[cfg(feature = "keyring")]
use crate::secrets::backend::SecretsBackend;
#[cfg(feature = "keyring")]
use crate::secrets::errors::{Result, SecretError};
#[cfg(feature = "keyring")]
use crate::secrets::types::{Scope, SecretKey, SecretValue};

#[cfg(feature = "keyring")]
#[derive(Debug, Clone)]
pub struct KeyringBackend {
    service_prefix: String,
}

#[cfg(feature = "keyring")]
impl KeyringBackend {
    pub fn new(service_prefix: impl Into<String>) -> Self {
        Self {
            service_prefix: service_prefix.into(),
        }
    }
    fn service(&self, scope: &Scope) -> String {
        format!("{}:{}", self.service_prefix, scope.label())
    }
}

#[cfg(feature = "keyring")]
impl SecretsBackend for KeyringBackend {
    fn set(&self, scope: &Scope, key: &SecretKey, value: &SecretValue) -> Result<()> {
        let service = self.service(scope);
        keyring::Entry::new(&service, &key.0)
            .and_then(|e| e.set_password(&value.0))
            .map_err(|e| SecretError::Backend(e.to_string()))
    }

    fn get(&self, scope: &Scope, key: &SecretKey) -> Result<String> {
        let service = self.service(scope);
        match keyring::Entry::new(&service, &key.0).and_then(|e| e.get_password()) {
            Ok(v) => Ok(v),
            Err(_) => Err(SecretError::NotFound {
                scope: scope.label().to_string(),
                key: key.0.clone(),
            }),
        }
    }

    fn list(&self, _scope: Option<&Scope>) -> Result<HashMap<(Option<String>, String), String>> {
        // Keyring crate does not support listing. Return empty.
        Ok(HashMap::new())
    }
}
